File: sdk/dart/dovahlink_client/lib/src/protocol/error_payload.dart
import 'package:json_annotation/json_annotation.dart';

import 'package:dovahlink_client_sdk/src/protocol/json_map.dart';
import 'package:dovahlink_client_sdk/src/protocol/protocol_format_exception.dart';
import 'package:dovahlink_client_sdk/src/shared/enums.dart';

part 'error_payload.g.dart';

/// Incoming `error` payload (`protocol/schema/README.md`'s `error`). Decode-only: the client
/// never sends `error`.
@JsonSerializable(checked: true, createToJson: false)
class ErrorPayload {
  /// Creates an error payload.
  const ErrorPayload({
    required this.code,
    required this.message,
    required this.retryable,
    this.details,
  });

  /// Decodes and validates one `error` payload.
  factory ErrorPayload.fromJson(JsonMap json) {
    try {
      return _$ErrorPayloadFromJson(json);
    } on Object catch (error) {
      throw ProtocolFormatException('Invalid error payload: $error');
    }
  }

  /// The canonical machine-readable failure code. For branching; never [message].
  @JsonKey(required: true)
  final ProtocolErrorCode code;

  /// Diagnostic text. Never used for branching.
  @JsonKey(required: true)
  final String message;

  /// Whether a fresh connection may retry.
  @JsonKey(required: true)
  final bool retryable;

  /// Safe diagnostic details, when any exist. Unlike the other fields, this key may be absent
  /// entirely, not only `null` -- per the schema's "optional when no safe diagnostic details
  /// exist."
  final JsonMap? details;
}

File: integration/DovahLinkValidationClient/ErrorPayload.cs
using System.Text.Json.Nodes;

namespace DovahLinkValidationClient;

/// <summary>
/// Incoming <c>error</c> payload (<c>protocol/schema/README.md</c>'s <c>error</c>). Decode-only:
/// this client never sends <c>error</c>.
/// </summary>
/// <param name="Code">The stable, machine-readable error code. For branching; never
/// <paramref name="Message"/>.</param>
/// <param name="Message">The human-readable, diagnostic-only description. Must not be used for
/// branching.</param>
/// <param name="Retryable">Whether retrying the failed operation is allowed.</param>
/// <param name="Details">Optional structured error details. Absent (the key may be omitted
/// entirely) when no safe diagnostic details exist.</param>
public sealed record ErrorPayload(string Code, string Message, bool Retryable, JsonNode? Details)
{
    /// <summary>
    /// Decodes and validates one <c>error</c> payload.
    /// </summary>
    /// <param name="payload">The envelope's decoded payload object.</param>
    /// <returns>The decoded error payload.</returns>
    /// <exception cref="FormatException">Thrown when a required field is missing or the wrong JSON
    /// type.</exception>
    public static ErrorPayload Decode(JsonObject payload)
    {
        try
        {
            string code = payload["code"]?.GetValue<string>() ?? throw new FormatException("Missing code.");
            string message = payload["message"]?.GetValue<string>() ?? throw new FormatException("Missing message.");
            bool retryable = payload["retryable"]?.GetValue<bool>() ?? throw new FormatException("Missing retryable.");
            JsonNode? details = payload["details"];
            return new ErrorPayload(code, message, retryable, details);
        }
        catch (InvalidOperationException ex)
        {
            throw new FormatException($"Malformed error payload: {ex.Message}", ex);
        }
    }
}
